lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -1600,27 +1600,77 @@ class ConvertAtenScaledDotProductAttentionOp
                                            "only default scale supported");
     }
 
+    auto opTy = cast<ValueTensorType>(op.getType()).toBuiltinTensor();
+    auto query = adaptor.getQuery();
+    auto value = adaptor.getValue();
+    auto key = adaptor.getKey();
+
+    auto loc = op.getLoc();
+    auto collapseBatch = [&rewriter, loc](Value value) -> Value {
+      auto valueTy = cast<ShapedType>(value.getType());
+      if (valueTy.getRank() <= 3)
+        return value;
+
+      llvm::SmallVector<int64_t, 3> newShape(3, 1);
+      llvm::SmallVector<ReassociationIndices, 3> reassociation(3);
+      reassociation[1].push_back(valueTy.getRank() - 2);
+      reassociation[2].push_back(valueTy.getRank() - 1);
+      newShape[1] = valueTy.getDimSize(valueTy.getRank() - 2);
+      newShape[2] = valueTy.getDimSize(valueTy.getRank() - 1);
+
+      for (int i = 0, s = valueTy.getRank() - 2; i < s; ++i) {
+        reassociation.front().push_back(i);
+        if (valueTy.isDynamicDim(i) || newShape[0] == ShapedType::kDynamic) {
+          newShape[0] = ShapedType::kDynamic;
+          continue;
+        }
+        newShape[0] = newShape[0] * valueTy.getDimSize(i);
```

Comment:
If we have a 4-d rank tensor, whose shape is <?x4x1024x64> then newShape[0] will be -4 ? Is this intended?

Reply:
Nope, check the bailout above. Once it goes dynamic it bails out.